libs/beacon/src/beacon_service.cpp

```diff
@@ -484,99 +270,93 @@ BeaconService::State BeaconService::OnVerifySignaturesState()
-                   " expected: ", current_entropy_.round);
-    return State::COLLECT_SIGNATURES;
-  }
+    uint64_t const index = block_entropy_being_created_->block_number;
 
-  // Success - Add relevant info
-  auto &signatures_struct = signatures_being_built_[current_entropy_.round];
-  auto &all_sigs_map      = signatures_struct.threshold_signatures;
+    if (ret.threshold_signatures.empty())
+    {
+      FETCH_LOG_DEBUG(LOGGING_NAME, "Peer wasn't ready when asking for signatures: ",
+                      qual_promise_identity_.identifier().ToBase64());
+      state_machine_->Delay(std::chrono::milliseconds(100));
 
-  for (auto const &address_sig_pair : ret.threshold_signatures)
-  {
-    all_sigs_map[address_sig_pair.first] = address_sig_pair.second;
-    // Let the manager know
-    AddSignature(address_sig_pair.second);
-  }
+      return State::COLLECT_SIGNATURES;
+    }
 
-  FETCH_LOG_INFO(LOGGING_NAME, "After adding, we have ", all_sigs_map.size(),
-                 " signatures. Round: ", current_entropy_.round);
+    if (ret.round != index)
+    {
+      FETCH_LOG_WARN(LOGGING_NAME,
+                     "Peer returned the wrong round when asked for signatures. Peer: ",
+                     qual_promise_identity_.identifier().ToBase64(), " returned: ", ret.round,
+                     " expected: ", index);
+      return State::COLLECT_SIGNATURES;
+    }
 
-  if (active_exe_unit_->manager.can_verify() && active_exe_unit_->manager.Verify())
-  {
-    // Storing the result of current entropy
-    current_entropy_.signature = active_exe_unit_->manager.GroupSignature();
-    auto sign                  = current_entropy_.signature.getStr();
-    current_entropy_.entropy   = crypto::Hash<crypto::SHA256>(crypto::Hash<crypto::SHA256>(sign));
+    // Success - Add relevant info
+    auto &signatures_struct = signatures_being_built_[index];
+    auto &all_sigs_map      = signatures_struct.threshold_signatures;
 
-    // Broadcasting the entropy to those listening, but not participating
-    if (broadcasting_)
+    for (auto const &address_sig_pair : ret.threshold_signatures)
     {
-      Serializer msgser;
-      msgser << current_entropy_;
-      endpoint_.Broadcast(SERVICE_DKG, CHANNEL_ENTROPY_DISTRIBUTION, msgser.data());
+      all_sigs_map[address_sig_pair.first] = address_sig_pair.second;
+      // Let the manager know
+      AddSignature(address_sig_pair.second);
+
+      // If we have collected enough signatures already then break
+      if (active_exe_unit_->manager.can_verify())
+      {
+        break;
+      }
     }
 
-    FETCH_LOG_INFO(LOGGING_NAME, "Generated new entropy value");
+    FETCH_LOG_INFO(LOGGING_NAME, "After adding, we have ", all_sigs_map.size(),
+                   " signatures. Round: ", index);
+  }  // Mutex unlocks here since verification can take some time
+
+  MilliTimer const timer{"Verify threshold signature", 100};
 
+  // TODO(HUT): possibility for infinite loop here I suspect.
```

Comment:
Issue?

Reply:
TODO 3 - partially jenny is workin on this